bitcoin/src/blockdata/block.rs

```diff
@@ -98,14 +99,57 @@ impl Decodable for Version {
 /// * [CBlock definition](https://github.com/bitcoin/bitcoin/blob/345457b542b6a980ccfbc868af0970a6f91d1b82/src/primitives/block.h#L62)
 #[derive(PartialEq, Eq, Clone, Debug)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-pub struct Block {
+pub struct Block<V = Unchecked>
+where
+    V: Validation,
+{
     /// The block header
     pub header: Header,
     /// List of transactions contained in the block
     pub txdata: Vec<Transaction>,
+    /// Validation marker.
+    marker: PhantomData<V>,
 }
 
-impl Block {
+impl Block<Unchecked> {
+    /// Constructs a new [`Block`].
+    ///
+    /// # Returns
+    ///
+    /// Return the block if it is valid, `None` if not. See [`Block::validate`].
+    pub fn new(header: Header, txdata: Vec<Transaction>) -> Option<Block<Checked>> {
+        let block = Block { header, txdata, marker: PhantomData::<Unchecked> };
+        block.validate()
+    }
+
+    /// Constructs a new `Block` without doing any validation.
+    pub fn new_unchecked(header: Header, txdata: Vec<Transaction>) -> Block<Unchecked> {
+        Block { header, txdata, marker: PhantomData::<Unchecked> }
+    }
+
+    /// Validates a block.
+    ///
+    /// Validation currently means:
+    ///
+    /// * The Merkle root of the header matches Merkle root of the transaction list.
+    /// * The witness commitment in coinbase matches the transaction list.
+    pub fn validate(self) -> Option<Block<Checked>> {
+        if self.check_merkle_root() && self.check_witness_commitment() {
+            Some(Block { header: self.header, txdata: self.txdata, marker: PhantomData::<Checked> })
+        } else {
+            None
+        }
+    }
+
+    /// Ignores block validation logic and just assumes you know what you are doing.
+    ///
+    /// You should only use this function if you trust the block i.e., it comes from a trusted node.
+    pub fn assume_checked(self) -> Block<Checked> {
+        Block { header: self.header, txdata: self.txdata, marker: PhantomData::<Checked> }
+    }
+}
+
+impl Block<Unchecked> {
     /// Returns the block hash.
     pub fn block_hash(&self) -> BlockHash { self.header.block_hash() }
```

Comment:
In a867ba36b12b640bd65fdccac32e4796aab79d02:

This should be available for an arbitrary `Block<V>`. And the `check_*` functions below should be made private.

Reply:
And we should cache `witness_root` in `Block<Checked>`. (This is "just an optimization" and could be deferred, but I'd like to do it now since we should nail down how exactly we should cache data only for a checked block without leaking too much noise into the API.)

Reply:
If we put `block_hash` in both traits it saves defining a third one just for it - I did that.

Reply:
`block_hash` doesn't need to be in an extension trait. The method is pretty core to what a block is.